File: code/packages/typescript/arithmetic/src/adders.ts
/**
 * Adder circuits built from logic gates.
 *
 * # Moving from Logic to Math
 *
 * In the logic-gates package, we saw how transistors combine to form gates
 * that perform basic Boolean operations (AND, OR, XOR). But how do we get a
 * computer to do actual math?
 *
 * This module answers that question. By creatively wiring together those
 * fundamental logic gates, we can build circuits that add binary numbers.
 * From a simple "Half Adder" that adds two individual bits, we build up to
 * a "Ripple Carry Adder" that adds multi-bit numbers — the same way you
 * add decimal numbers on paper, column by column, carrying as you go.
 *
 * Half adder: adds two bits.
 * Full adder: adds two bits + carry-in.
 * Ripple carry adder: chains full adders for N-bit addition.
 */

import { AND, OR, XOR, type Bit } from "@coding-adventures/logic-gates";

/**
 * Add two single bits.
 *
 * # Why "Half"?
 *
 * Adding two binary bits is simple, but we have to account for carrying over
 * to the next column — exactly like grade-school addition:
 *
 *       1
 *     + 1
 *     ---
 *      10  (which is 2 in binary)
 *
 * In the 1s column, the sum is 0, and we "carry" a 1 to the next column.
 * The Half Adder produces both these outputs: a Sum bit and a Carry bit.
 * It is called a "Half" adder because, while it can generate a carry, it
 * cannot ACCEPT a carry input from a previous column.
 *
 * Truth table:
 *
 *     A | B | Sum | Carry
 *     --|---|-----|------
 *     0 | 0 |  0  |   0
 *     0 | 1 |  1  |   0
 *     1 | 0 |  1  |   0
 *     1 | 1 |  0  |   1
 *
 * If you look closely at the truth table:
 * - Sum is exactly the XOR operation (1 only when inputs differ).
 * - Carry is exactly the AND operation (1 only when both inputs are 1).
 *
 * @param a - First bit (0 or 1)
 * @param b - Second bit (0 or 1)
 * @returns [sum, carry] — the sum bit and the carry bit
 */
export function halfAdder(a: Bit, b: Bit): [Bit, Bit] {
  const sumBit = XOR(a, b);
  const carry = AND(a, b);
  return [sumBit, carry];
}

/**
 * Add two bits plus a carry-in from a previous addition.
 *
 * # Handling the Ripple
 *
 * To add multi-bit numbers, every column beyond the first might receive a
 * carry from the column to its right. A Full Adder takes three inputs (A, B,
 * and CarryIn) and correctly produces a Sum and a CarryOut.
 *
 * Built from two half adders and an OR gate:
 *   1. Half-add A and B → partial_sum, partial_carry
 *   2. Half-add partial_sum and carry_in → sum, carry2
 *   3. carry_out = OR(partial_carry, carry2)
 *
 * If EITHER step generated a carry, our final CarryOut is 1.
 *
 * Truth table:
 *
 *     A | B | Cin | Sum | Cout
 *     --|---|-----|-----|-----
 *     0 | 0 |  0  |  0  |  0
 *     0 | 0 |  1  |  1  |  0
 *     0 | 1 |  0  |  1  |  0
 *     0 | 1 |  1  |  0  |  1
 *     1 | 0 |  0  |  1  |  0
 *     1 | 0 |  1  |  0  |  1
 *     1 | 1 |  0  |  0  |  1
 *     1 | 1 |  1  |  1  |  1
 *
 * @param a - First bit
 * @param b - Second bit
 * @param carryIn - Carry from previous column
 * @returns [sum, carryOut]
 */
export function fullAdder(a: Bit, b: Bit, carryIn: Bit): [Bit, Bit] {
  const [partialSum, partialCarry] = halfAdder(a, b);
  const [sumBit, carry2] = halfAdder(partialSum, carryIn);
  const carryOut = OR(partialCarry, carry2);
  return [sumBit, carryOut];
}

/**
 * Add two N-bit numbers using a chain of full adders.
 *
 * # The Ripple Effect
 *
 * Just like you add large numbers on paper starting from the rightmost digit
 * and moving left, the Ripple Carry Adder lines up a series of Full Adders.
 * The CarryOut of bit 0 is wired directly into the CarryIn of bit 1. The
 * CarryOut of bit 1 goes into bit 2, and so on.
 *
 * The worst-case performance is when adding something like 1111 + 0001. The
 * carry generated at the first bit must "ripple" all the way through every
 * single adder before the final sum is ready. In physical hardware, this
 * takes time, which is why modern CPUs use faster tricks like "Carry Lookahead".
 *
 * ```
 *      A3 B3      A2 B2      A1 B1      A0 B0
 *       |  |       |  |       |  |       |  |
 *     +----+     +----+     +----+     +----+
 *     | FA |<----| FA |<----| FA |<----| FA |<-- 0 (initial carry)
 *     +----+     +----+     +----+     +----+
 *        S3         S2         S1         S0
 * ```
 *
 * @param a - First number as array of bits, LSB first (index 0 = least significant).
 * @param b - Second number as array of bits, LSB first.
 * @param carryIn - Initial carry (default 0).
 * @returns [sumBits, carryOut] where sumBits is LSB first.
 *
 * @example
 * ```ts
 * // 5 + 3 = 8
 * const a: Bit[] = [1, 0, 1, 0];  // 5 in binary (LSB first: 1*1 + 0*2 + 1*4 + 0*8)
 * const b: Bit[] = [1, 1, 0, 0];  // 3 in binary (LSB first: 1*1 + 1*2 + 0*4 + 0*8)
 * const [result, carry] = rippleCarryAdder(a, b);
 * // result = [0, 0, 0, 1], carry = 0  → 8 in binary
 * ```
 */
export function rippleCarryAdder(
  a: Bit[],
  b: Bit[],
  carryIn: Bit = 0
): [Bit[], Bit] {
  if (a.length !== b.length) {
    throw new Error(
      `a and b must have the same length, got ${a.length} and ${b.length}`
    );
  }
  if (a.length === 0) {
    throw new Error("bit lists must not be empty");
  }

  const sumBits: Bit[] = [];
  let carry: Bit = carryIn;

  for (let i = 0; i < a.length; i++) {
    const [sumBit, newCarry] = fullAdder(a[i], b[i], carry);
    sumBits.push(sumBit);
    carry = newCarry;
  }

  return [sumBits, carry];
}

File: code/packages/rust/arithmetic/src/adders.rs
//! Adder circuits built from logic gates.
//!
//! # From gates to addition
//!
//! How do you add two numbers using only AND, OR, and XOR gates? The key
//! insight is that binary addition of two single bits works exactly like
//! XOR (for the sum) and AND (for the carry):
//!
//! ```text
//! 0 + 0 = 0, carry 0   ->  XOR(0,0)=0, AND(0,0)=0
//! 0 + 1 = 1, carry 0   ->  XOR(0,1)=1, AND(0,1)=0
//! 1 + 0 = 1, carry 0   ->  XOR(1,0)=1, AND(1,0)=0
//! 1 + 1 = 0, carry 1   ->  XOR(1,1)=0, AND(1,1)=1
//! ```
//!
//! This is the **half adder**. To handle carry-in from a previous column,
//! we chain two half adders into a **full adder**. To add N-bit numbers,
//! we chain N full adders into a **ripple-carry adder**.
//!
//! # Why "ripple carry"?
//!
//! The carry "ripples" from the least significant bit to the most significant
//! bit, one full adder at a time. This is simple but slow for large bit
//! widths because each adder must wait for the previous one's carry output.
//! Real CPUs use faster designs (carry-lookahead, carry-select) but they
//! all build on the same fundamental principles shown here.

use logic_gates::gates::{and_gate, or_gate, xor_gate};

/// Half adder — adds two single bits.
///
/// Returns `(sum, carry)`.
///
/// The half adder is the simplest adding circuit. It handles the addition
/// of two bits but cannot accept a carry-in from a previous addition.
///
/// # Circuit
///
/// ```text
/// a --+--[XOR]-- sum
///     |
/// b --+--[AND]-- carry
/// ```
///
/// # Truth table
///
/// ```text
/// A  B  | Sum  Carry
/// ------+-----------
/// 0  0  |  0     0     (0 + 0 = 0)
/// 0  1  |  1     0     (0 + 1 = 1)
/// 1  0  |  1     0     (1 + 0 = 1)
/// 1  1  |  0     1     (1 + 1 = 10 in binary)
/// ```
///
/// # Example
///
/// ```
/// use arithmetic::adders::half_adder;
/// assert_eq!(half_adder(1, 1), (0, 1)); // 1 + 1 = 10 binary
/// assert_eq!(half_adder(1, 0), (1, 0)); // 1 + 0 = 1
/// ```
#[inline]
pub fn half_adder(a: u8, b: u8) -> (u8, u8) {
    (xor_gate(a, b), and_gate(a, b))
}

/// Full adder — adds two bits plus a carry-in from a previous addition.
///
/// Returns `(sum, carry_out)`.
///
/// Built from two half adders and an OR gate:
/// 1. Half-add `a` and `b` -> `partial_sum`, `partial_carry`
/// 2. Half-add `partial_sum` and `carry_in` -> `sum`, `carry2`
/// 3. `carry_out = OR(partial_carry, carry2)`
///
/// # Circuit
///
/// ```text
/// a --------+
///           |--[HA]-- partial_sum --+
/// b --------+         |             |--[HA]-- sum
///                     |  carry_in --+         |
///                     |                       |
///           partial_carry            carry2   |
///                     |                |      |
///                     +--[OR]----------+      |
///                         |                   |
///                     carry_out                |
/// ```
///
/// # Example
///
/// ```
/// use arithmetic::adders::full_adder;
/// assert_eq!(full_adder(1, 1, 0), (0, 1)); // 1+1+0 = 10
/// assert_eq!(full_adder(1, 1, 1), (1, 1)); // 1+1+1 = 11
/// ```
#[inline]
pub fn full_adder(a: u8, b: u8, carry_in: u8) -> (u8, u8) {
    let (partial_sum, partial_carry) = half_adder(a, b);
    let (sum, carry2) = half_adder(partial_sum, carry_in);
    let carry_out = or_gate(partial_carry, carry2);
    (sum, carry_out)
}

/// Result of a ripple-carry addition.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RippleCarryResult {
    /// Sum bits (LSB first, index 0 = least significant bit).
    pub sum: Vec<u8>,
    /// Carry out of the most significant bit position.
    pub carry_out: u8,
    /// Signed overflow: set when adding two positive numbers produces a
    /// negative result, or two negative numbers produces positive.
    /// Only meaningful when interpreting bits as two's complement.
    pub overflow: bool,
}

/// Ripple-carry adder — adds two N-bit numbers using a chain of full adders.
///
/// Each full adder handles one bit position, passing its carry output to
/// the next position. The carry "ripples" from LSB to MSB.
///
/// # Arguments
///
/// - `a` — first number as bits, LSB first (index 0 = least significant)
/// - `b` — second number as bits, LSB first
///
/// Both slices must have the same length and be non-empty.
///
/// # Carry-in
///
/// The initial carry-in defaults to 0. For subtraction (A - B), you'd
/// invert B and set carry-in to 1 (two's complement negation). The ALU
/// module handles this automatically.
///
/// # Example
///
/// ```
/// use arithmetic::adders::ripple_carry_adder;
///
/// // 5 + 3 = 8
/// // 5 = 0101 (LSB first: [1, 0, 1, 0])
/// // 3 = 0011 (LSB first: [1, 1, 0, 0])
/// // 8 = 1000 (LSB first: [0, 0, 0, 1])
/// let result = ripple_carry_adder(&[1, 0, 1, 0], &[1, 1, 0, 0]);
/// assert_eq!(result.sum, vec![0, 0, 0, 1]);
/// assert_eq!(result.carry_out, 0);
/// ```
pub fn ripple_carry_adder(a: &[u8], b: &[u8]) -> RippleCarryResult {
    assert_eq!(
        a.len(),
        b.len(),
        "a and b must have the same length, got {} and {}",
        a.len(),
        b.len()
    );
    assert!(!a.is_empty(), "bit lists must not be empty");

    let mut sum = Vec::with_capacity(a.len());
    let mut carry: u8 = 0;

    for i in 0..a.len() {
        let (s, c) = full_adder(a[i], b[i], carry);
        sum.push(s);
        carry = c;
    }

    // Signed overflow detection: overflow occurs when the carry INTO the
    // sign bit differs from the carry OUT of the sign bit. Equivalently,
    // when adding two positive numbers gives negative, or vice versa.
    let n = a.len();
    let a_sign = a[n - 1];
    let b_sign = b[n - 1];
    let result_sign = sum[n - 1];
    let overflow = (a_sign == b_sign) && (result_sign != a_sign);

    RippleCarryResult {
        sum,
        carry_out: carry,
        overflow,
    }
}

/// Ripple-carry adder with explicit carry-in, used internally by the ALU
/// for subtraction (where carry_in = 1).
pub fn ripple_carry_adder_with_carry(a: &[u8], b: &[u8], carry_in: u8) -> RippleCarryResult {
    assert_eq!(
        a.len(),
        b.len(),
        "a and b must have the same length, got {} and {}",
        a.len(),
        b.len()
    );
    assert!(!a.is_empty(), "bit lists must not be empty");

    let mut sum = Vec::with_capacity(a.len());
    let mut carry = carry_in;

    for i in 0..a.len() {
        let (s, c) = full_adder(a[i], b[i], carry);
        sum.push(s);
        carry = c;
    }

    let n = a.len();
    let a_sign = a[n - 1];
    let b_sign = b[n - 1];
    let result_sign = sum[n - 1];
    let overflow = (a_sign == b_sign) && (result_sign != a_sign);

    RippleCarryResult {
        sum,
        carry_out: carry,
        overflow,
    }
}

// ===========================================================================
// Inline unit tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_half_adder_truth_table() {
        assert_eq!(half_adder(0, 0), (0, 0));
        assert_eq!(half_adder(0, 1), (1, 0));
        assert_eq!(half_adder(1, 0), (1, 0));
        assert_eq!(half_adder(1, 1), (0, 1));
    }

    #[test]
    fn test_full_adder_truth_table() {
        // All 8 combinations of (a, b, carry_in)
        assert_eq!(full_adder(0, 0, 0), (0, 0)); // 0+0+0 = 0
        assert_eq!(full_adder(0, 0, 1), (1, 0)); // 0+0+1 = 1
        assert_eq!(full_adder(0, 1, 0), (1, 0)); // 0+1+0 = 1
        assert_eq!(full_adder(0, 1, 1), (0, 1)); // 0+1+1 = 10
        assert_eq!(full_adder(1, 0, 0), (1, 0)); // 1+0+0 = 1
        assert_eq!(full_adder(1, 0, 1), (0, 1)); // 1+0+1 = 10
        assert_eq!(full_adder(1, 1, 0), (0, 1)); // 1+1+0 = 10
        assert_eq!(full_adder(1, 1, 1), (1, 1)); // 1+1+1 = 11
    }

    #[test]
    fn test_ripple_carry_5_plus_3() {
        // 5 + 3 = 8 in 4-bit
        // In signed 4-bit two's complement, range is -8 to 7.
        // 5 + 3 = 8 overflows because the result (1000) has MSB=1
        // while both inputs had MSB=0 (positive + positive = negative).
        let result = ripple_carry_adder(&[1, 0, 1, 0], &[1, 1, 0, 0]);
        assert_eq!(result.sum, vec![0, 0, 0, 1]); // 8 in LSB-first
        assert_eq!(result.carry_out, 0);
        assert!(result.overflow); // Signed overflow in 4-bit
    }

    #[test]
    fn test_ripple_carry_overflow() {
        // 255 + 1 in 8-bit: carries out
        let a = vec![1, 1, 1, 1, 1, 1, 1, 1]; // 255
        let b = vec![1, 0, 0, 0, 0, 0, 0, 0]; // 1
        let result = ripple_carry_adder(&a, &b);
        assert_eq!(result.sum, vec![0, 0, 0, 0, 0, 0, 0, 0]); // 0
        assert_eq!(result.carry_out, 1);
    }
}
